Met.process.updates

rykelly opened the pull request:
This one's a doozy... Submitting as a pull request so we can discuss easily (I like the line comments feature and think this is the only way?), but this might be something to keep as a branch for @tonygardella or others to maybe get a few ideas from. I had to guess at how I think some things should work, almost certainly broke browndog since I'm not using it and know nothing about it, etc.

On the upside, at least for my use case, this branch now (1) allows processed met to be overwritten as we've discussed, and (2) will properly extend an existing input record to accommodate requested start/end dates, rather than creating a new one and, therefore, a big mess of conflicting input and dbfile records. 

I'm out of town from tomorrow through Monday but will try to be responsive if possible (not sure about internet situation yet).

rykelly commented on db/R/dbfiles.R:
```diff
@@ -65,12 +66,53 @@ dbfile.input.insert <- function(in.path, in.prefix, siteid, startdate, enddate,
+    
+    if(existing.input$start_date != startdate || existing.input$end_date != enddate) {
+      print(existing.input, digits=10)
+      logger.error(paste0(
```
Most of the changes in dbfiles.R, aside from cosmetics, added errors like this one. I know it might not be the intent to be so strict, in which case abandon these. But I would say that the whole inputs/dbfiles thing could be a lot less confusing if it were more rigidly defined.

rykelly replied:
I second that emotion. Isn't this something that can be hard-coded into the PSQL schema? Then there's an extra level of security. Just a thought (brought to mind while I was working on this, and also fighting with BETY to let me delete erroneous records—she clearly knows things like you can't orphan a record with `parent_id`, etc.)

rykelly commented on db/R/utils.R:
```diff
@@ -107,17 +107,17 @@ db.open <- function(params) {
 ##' db.close(con)
 ##' }
-db.close <- function(con) {
+db.close <- function(con, showWarnings=TRUE) {
```
This is related to adding a `on.exit()` statement to met.process ([here](https://github.com/PecanProject/pecan/pull/1016/files#diff-a796467dc73dfb5ca85a17b225f8f409R23). I think that should be used in more places so that connections are always closed, even if there are errors. But if you do that but someone's explicitly closed the connection in the meantime you get a warning, so I added this option to suppress it.

rykelly commented on modules/data.atmosphere/R/download.raw.R:
```diff
@@ -1,45 +0,0 @@
-download.raw <- function(data.set,outfolder,pkg,raw.host, ...){
```
This function wasn't being used anywhere.

rykelly commented on modules/data.atmosphere/R/download.raw.met.module.R:
```diff
@@ -0,0 +1,64 @@
+.download.raw.met.module <- function(dir, met, register, machine, start_date, end_date, con, new.site=NULL, host, overwrite=FALSE) {
```
This one, along with the other *.module files, simply moved big chunks of code out of met.process. I kind of had @mdietze's #1009 idea in mind, but this is as far as I got. In any case, it's definitely saved my scroll-wheel finger some miles.

rykelly commented on modules/data.atmosphere/R/listToArgString.r:
```diff
@@ -0,0 +1,43 @@
+  return(arg.string)
+}
+
+.parseArg <- function(x) {
```
As usual, all my fns need better documentation. One note for now though, the ones that start with `.` are not meant to be exported. I think I read that's an R convention somewhere.

rykelly commented on modules/data.atmosphere/R/met.process.R:
```diff
@@ -10,21 +10,23 @@
-met.process <- function(site, input_met, start_date, end_date, model, host, dbparms, dir, browndog=NULL){
-# browser()
+##' @author Elizabeth Cowdery, Michael Dietze, Ankur Desai, James Simkins, Ryan Kelly
+met.process <- function(
```
Most of what's changed here is formatting and the above mentioned move of "modules" into their own functions. No real functionality changes, IIRC.

rykelly commented on utils/R/convert.input.R:
```diff
@@ -6,65 +6,144 @@
+      # start and end dates stay the same. 
+      # collect files to flag for deletion
+      existing.dbfile <- dbfile.check('Input', existing.input[['id']], con, host$name)
+      if(nrow(existing.dbfile) > 0) {
```
By this point convert.input knows there is an existing input record and dbfile record, and overwrite is TRUE. So there are going to be files to delete. But it needs to wait until it know it's been successful. So, here it collects the file paths to be deleted, and creates string versions of the remote functions needed to (1) move them to a temp location, (2) delete the temp location (if successful) and/or (3) replace them to their original locations (if unsuccessful.

Some limitations:
1. If convert.input is partially successful, you could end up with some new (but presumably unwanted) files, mixed in with the old ones that were replaced on exiting this failing function.
2. I don't actually check to make sure that the remote function successfully executes. I think this is probably ok in most places, because if it doesn't return a `result` object with a valid `result$file` then this function should error out, triggering the replace-old-files routine. But that needs some testing.
3. I meant to but haven't added a mechanism to delete unused dbfile entries. Normally I think the overwriting call to convert.inputs will be reusing the same input and dbfile records, so it's no problem. But for the case where you have a dbfile record pointing to one location, and now decide to write to another, we may want to make sure to obliterate the obsolete one?

rykelly commented on utils/R/convert.input.R:
```diff
@@ -6,65 +6,144 @@
+      logger.info("Skipping this input conversion because files are already available.")
+      return(list(input.id=existing.input$id, dbfile.id=existing.dbfile$id))
+    } else {
+      # Start/end dates need to be updated so that the input spans a continuous timeframe
```
If I understood correctly, this is the desired behavior: If there's any matching input record already (minus dates), then it should be extended to accommodate its entire current range and the newly requested start/end range. Anyway, that's what this version does. Sucks to be you if 2000-2001 exists and now you want to run 2015-2016, because you have to wait for all intervening years to complete...

rykelly commented on utils/R/convert.input.R:
```diff
@@ -6,65 +6,144 @@
+    return(NULL)
   }
 
+  if(missing(input.id) || is.na(input.id) || is.null(input.id)) {
```
Here's one of just a few places where it actually matters whether convert.inputs was passed an input.id (as mentioned, download.raw.met now uses this too, but there's no input for that one of course).

rykelly commented on utils/R/convert.input.R:
```diff
@@ -6,65 +6,144 @@
   conversion = "local.remote" #default
   
-  if(!is.null(browndog) & host$name == 'localhost'){ # perform conversions with Brown Dog - only works locally right now
+  if(!is.null(browndog) && host$name == 'localhost'){ 
```
To reiterate, I did *nothing* functional to code within browndog-only blocks. Apparently I obsessive-compulsively did some formatting, though.

rykelly commented on utils/R/convert.input.R:
```diff
@@ -138,57 +217,100 @@ convert.input <- function(input.id, outfolder, formatname, mimetype, site.id, st
+    fcn.args$start_date <- start_date
+    fcn.args$end_date <- end_date 
+
+    arg.string <- listToArgString(fcn.args)
```
Here's the new approach to handling argument passing. Start with the `...` args (called input.args here), which are already a named list. Then add any other args that you know you want, with the desired names. Then pass them to `listToArgString` to get a string version you can paste into the remote call. This is pretty flexible, and importantly since all arguments are named it lets you omit them willy-nilly as long as whatever fcn you're calling accepts `...` (so it doesn't bonk out when receiving arguments it can't use) and you don't omit anything that it actually requires. 
So for example, the download.* functions still receive `in.path` and `in.prefix` args (which will be set to NULL since no dbfile will have been found), but they pass through `...` unnoticed since they're not used.

rykelly commented on utils/R/convert.input.R:
```diff
@@ -138,57 +217,100 @@ convert.input <- function(input.id, outfolder, formatname, mimetype, site.id, st
   ## insert new record into database
-  if(write==TRUE){
+  if(write==TRUE) {
+    if(exists("existing.input") && nrow(existing.input) > 0 &&
```
If everything went smoothly and there are indeed new files to be registered and there was a matching input already, here's where we update its start/end dates. Didn't do it earlier because if something goes wrong you don't want the input record to be altered.

rykelly commented on utils/R/convert.input.R:
```diff
@@ -138,57 +217,100 @@ convert.input <- function(input.id, outfolder, formatname, mimetype, site.id, st
-  if(write==TRUE){
+  if(write==TRUE) {
+    if(exists("existing.input") && nrow(existing.input) > 0 &&
+       (existing.input$start_date != start_date || existing.input$end_date != end_date)) {
```
This check means that even if you've overwritten files, the input record won't be updated if the start/end dates haven't been altered. There's probably an argument for doing the update anyway so `updated_at` gets updated.

rykelly commented on utils/R/convert.input.R:
```diff
@@ -138,57 +217,100 @@ convert.input <- function(input.id, outfolder, formatname, mimetype, site.id, st
+        "updated_at=NOW() WHERE id=", existing.input$id), con)
+    }
+
+    parent.id <- ifelse(is.null(input), NA, input$id)
```
Another little tweak for the special case that there is no starting input.

rykelly commented on utils/R/convert.input.R:
```diff
@@ -138,57 +217,100 @@ convert.input <- function(input.id, outfolder, formatname, mimetype, site.id, st
                                     con = con,
                                     hostname = machine$hostname) 
+
+    successful <- TRUE
```
See above for the two alternative functions that will be run by `on.exit` depending on whether we make it all the way here and successful==TRUE.

mdietze commented on modules/data.atmosphere/R/met2model.module.R:
```diff
@@ -0,0 +1,43 @@
+    model.id = ready.id
+    
+    if("CRUNCEP" %in% met){outfolder <- file.path(dir,paste0(met,"_site_",str_ns))}
+    if("GFDL" %in% met){outfolder <- file.path(dir,paste0(met,"_site_",str_ns))}
```
These hacks need to be collapsed (they're identical) and in general (throughout met.process) replaced with a registry tag identifying all cases where met is directly extracted (the number of products in this category is growing rapidly)

mdietze commented on utils/R/convert.input.R:
```diff
@@ -6,65 +6,133 @@
+      remote.execute.R(file.deletion.commands$move.to.tmp, host, user=NA, verbose=TRUE, R="R")
+      
+      # Schedule files to be replaced or deleted on exiting the function
+      succesful <- FALSE
```
typo

rykelly replied:
Shit, thanks. This explains a bug I was hitting for awhile and, apparently, fixed some incorrect way.

mdietze commented on utils/R/convert.input.R:
```diff
@@ -6,65 +6,133 @@
+      )
+    } else if((start_date >= existing.input$start_date) && (end_date <= existing.input$end_date)) {
+      # There's an existing input that spans desired start/end dates. Use that one. 
+      logger.info("Skipping this input conversion because files are already available.")
```
Ok, so where things get complicated is with met2model. The registry will identify two cases: models that can read met by date and those that can't. For the former, this is the desired behavior -- we can just read the existing file and ignore the parts we don't need. For the latter, we need to create a new met file that just contains the dates we need and insert it into Inputs/dbfiles as a new entry (i.e. we can't just use overwrite=TRUE to force it to make a new file).  Probably what we need is an additional convert.inputs argument to handle this case. Thankfully, like overwrite, it's a fairly simple case -- unless the dates are an exact match you always convert.

mdietze commented:
Overall looks like a nice step forward. Requested a few minor changes that shouldn't take long to resolve. Need to also resolve conflicts and figure out why Travis is unhappy.  FYI the flags to get rid of the CRUNCEP and GFDL hacks are in @tonygardella's PR as ```<direct.extract>```, so it's ok to punt on those and focus on if/how the modules should be split up better.  I'm leaning toward accepting this PR and then having Tony address merging any remaining logic from his changes into this and test everything (sorry Tony), as I think this approach looks cleaner.

tonygardella commented:
@rykelly @mdietze I'm in the midst of cleaning my PR up and fixing it to fit Ryan's changes so no problem.

rykelly commented on modules/data.atmosphere/R/download.raw.met.module.R:
```diff
@@ -0,0 +1,29 @@
+.download.raw.met.module <- function(dir, met, register, machine, start_date, end_date, con, 
```
Only because I envisioned that the generic super-module would be the interface and details would be hidden away. But now that you mention it I could see someone just wanting to use this function rather than figure out where the desired dataset is hidden on the web...

rykelly commented on modules/data.atmosphere/R/extract.nc.module.R:
```diff
@@ -0,0 +1,45 @@
+      overwrite=overwrite)
+    
+  } else if(register$scale=="site") { ##### Site Level Processing
+    logger.info("Gapfilling") # Does NOT take place on browndog!
```
Gotcha. I really didn't pay much attention, but these were together (in some sense—maybe just comment-delimited) in met.process. Makes sense to separate.

rykelly replied:
Now that I look at this, I really don't understand... Why is the option either to extract (if register$scale==region) *or* to gapfill (if register$scale==site)?

mdietze replied:
Well, typically regional products have typically been gapfilled already and site-level data typically doesn't have to be extracted.

More generally, if we go down the route I proposed then we'll move toward building the workflow more dynamically rather than either/or

rykelly replied:
OK, split these into separate fns, and move the `if` statement to met.process (but didn't change how it works)

rykelly commented:
@mdietze , @tonygardella , sounds good. I'll try to tackle the issues Mike raised in the next few days. I have some travel time (en route to macrosystems meeting) that might be good for this. I think the Travis issues are the same ones from my other PR, which should be easy fixes. 

The nice thing about this approach is that it relies on convert.inputs at every step to do the checking / handle overwriting in a consistent manner. And thankfully @bcow wrote all her underlying functions with an overwrite option, so the convert.inputs logic isn't even that complicated. Definitely room for improvement but it's fairly clean and has been working well for me lately.

rykelly commented on modules/data.atmosphere/R/download.raw.met.module.R:
```diff
@@ -0,0 +1,29 @@
+  if(register$scale=="regional") { 
+    # Right now this only means NARR but will need to be generalized once we have more 
+    # regional met products
+     raw.data.site.id <- ifelse(met %in% c("CRUNCEP", "GFDL"), new.site$id, register$siteid)
```
OK, moved all hacks up to the top-level met.process function as suggested. I think you're right that they need to be reworked to be less hacky (read from the register, and/or query the db). But I'm really not the right person to tackle that, at least with the very limited hrs I've got left to devote on the current project. Let's figure out some funds to hire me for more PEcAn refactoring!

rykelly commented on modules/data.atmosphere/R/met.process.R:
```diff
@@ -41,328 +83,98 @@ met.process <- function(site, input_met, start_date, end_date, model, host, dbpa
+      overwrite=overwrite$download
+    )
+    if(met %in% c("CRUNCEP", "GFDL")) {
+      ready.id <- raw.id
```
OK, doing that now. This is one of those things that met.process should figure out from the registry though, I think.

rykelly commented on modules/data.atmosphere/R/met2cf.module.R:
```diff
@@ -0,0 +1,73 @@
+  format.id <- 33
+  
+  # Probably are other formats for which the CF file is not site specific?
+  new.site.id <- ifelse(met %in% c("NARR"), register$siteid, site$id) 
```
This is one of those things I don't really know about. Moved the check to met.process but otherwise left alone.

rykelly commented:
Latest push addresses most of the issues @mdietze raises above. Exception is that comment about sometimes (for models that need exact start/end dates) *not* reusing an input record with start/end dates spanning (but not identical to) the desired range.
